jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -1027,89 +1069,94 @@ public void run()
+                                long shrinkThreshold = _shrinkThreshold.get();
+                                long threshold = shrinkThreshold;
+
+                                // If the threshold is too far in the past,
+                                // advance it to be one idle timeout before now plus a bit of shrink period.
+                                if (NanoTime.elapsed(threshold, now) > idleTimeoutNanos)
+                                    threshold = now - idleTimeoutNanos;
+
+                                // advance the threshold by one shrink period
+                                threshold += shrinkPeriod;
+
+                                // Is the new threshold in the future?
+                                if (NanoTime.isBefore(now, threshold))
                                 {
+                                    // Yes - we cannot shrink yet, so break and continue looking for jobs
                                     if (LOG.isDebugEnabled())
-                                        LOG.debug("shrinking {}", QueuedThreadPool.this);
+                                        LOG.debug("Shrink skipped, threshold={}ms in future {}", NanoTime.millisElapsed(now, threshold), QueuedThreadPool.this);
                                     break;
                                 }
-                            }
 
-                            // Wait for a job, only after we have checked if we should shrink
-                            job = idleJobPoll(idleTimeout);
-
-                            // If still no job?
-                            if (job == null)
-                                // continue to try again
-                                continue;
+                                // We can shrink if we can update the atomic shrink threshold?
+                                if (_shrinkThreshold.compareAndSet(shrinkThreshold, threshold))
+                                {
+                                    // Yes - we have shrunk, so exit.
+                                    if (LOG.isDebugEnabled())
+                                        LOG.debug("SHRUNK, threshold={}ms in past {}", NanoTime.millisElapsed(threshold, now), QueuedThreadPool.this);
+                                    break exit;
+                                }
+                            }
                         }
-
-                        idle = false;
-
-                        // run job
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("run {} in {}", job, QueuedThreadPool.this);
-                        runJob(job);
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("ran {} in {}", job, QueuedThreadPool.this);
                     }
                     catch (InterruptedException e)
                     {
+                        boolean interrupted = Thread.interrupted();
```

Comment:
Won't this virtually always be false as `InterruptedException` is supposed to be thrown after a `Thread.interrupted()` returns true?